apache/superset
# perf(dashboard): Virtualization POC

### SUMMARY
This PR is a proof of concept for dashboard virtualization.
Due to a performance issues on large dashboards (60+ charts in a single tab/view), this PR introduces virtualization in an attempt to reduce the number of elements (DOM nodes) rendered at once.
In the initial state, we render only the charts in current viewport and charts max 1 view height away from current viewport.
When user scrolls down, charts that cross the 1 view height margin get rendered too.
We remove charts from DOM tree when they are at least 4 view heights away from current viewport.
Those values were chosen arbitrarily/experimentally to provide good middle-ground between memory efficiency and good user experience when scrolling up/down.

To sum it up: if chart is not rendered and it's less than 1 view height away from viewport - render it; if chart is rendered and it's more than 4 view heights away from viewport - remove it from DOM

Rerendering charts do NOT trigger new queries - we reuse data fetched on dashboard mount and only redraw the charts.

As of now, this feature is in proof-of-concept faze and should be considered as experimental. It's hidden behind a feature flag `DASHBOARD_VIRTUALIZATION`

Virtualization is disabled for bot workers (like Celery) to ensure that reports and alerts work correctly

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


https://user-images.githubusercontent.com/15073128/194769147-c8060f15-6ea9-403b-aaf0-8270227b94a1.mov




### TESTING INSTRUCTIONS
0. Enable FF `DASHBOARD_VIRTUALIZATION`
1. Open a dashboard with a lot of charts
2. Scroll up and down - verify that only the charts in current viewport (or close to it - 100% margin) are rendered. The rest of them displays a spinner and loads only when user scroll them into view
4. Verify that charts render when user scrolls, but do not send new queries to the backend

### ADDITIONAL INFORMATION


- [x] Required feature flags: `DASHBOARD_VIRTUALIZATION`
